Mathlib/Algebra/GroupPower/Basic.lean

```diff
@@ -3,71 +3,454 @@ Copyright (c) 2015 Jeremy Avigad. All rights reserved.
+
+## Implementation details
+
+We adopt the convention that `0^0 = 1`.
+-/
+
+universe u v w x y z u₁ u₂
+
+variable {α : Type _} {M : Type u} {N : Type v} {G : Type w} {H : Type x} {A : Type y} {B : Type z}
+  {R : Type u₁} {S : Type u₂}
+
+/-!
+### Commutativity
+
+First we prove some facts about `semiconj_by` and `commute`. They do not require any theory about
+`pow` and/or `nsmul` and will be useful later in this file.
+-/
+
+
+section Pow
+
+variable [Pow M ℕ]
+
+@[simp]
+theorem pow_ite (P : Prop) [Decidable P] (a : M) (b c : ℕ) :
+    (a ^ if P then b else c) = if P then a ^ b else a ^ c := by split_ifs <;> rfl
+#align pow_ite pow_ite
+
+@[simp]
+theorem ite_pow (P : Prop) [Decidable P] (a b : M) (c : ℕ) :
+    (if P then a else b) ^ c = if P then a ^ c else b ^ c := by split_ifs <;> rfl
+#align ite_pow ite_pow
+
+end Pow
+
 
 /-!
 ### Monoids
 -/
 
-section AddMonoid
+section Monoid
 
-variable [AddMonoid M]
+variable [Monoid M] [AddMonoid A]
 
-theorem nsmul_zero (n : ℕ) : n • (0 : M) = 0 := by
+theorem nsmul_zero (n : ℕ) : n • (0 : A) = 0 := by
   induction' n with n ih
   · exact zero_nsmul _
   · rw [succ_nsmul, ih, add_zero]
 
 @[simp]
-theorem one_nsmul (a : M) : 1 • a = a := by rw [succ_nsmul, zero_nsmul, add_zero]
+theorem one_nsmul (a : A) : 1 • a = a := by rw [succ_nsmul, zero_nsmul, add_zero]
 
-theorem add_nsmul (a : M) (m n : ℕ) : (m + n) • a = m • a + n • a := by
+theorem add_nsmul (a : A) (m n : ℕ) : (m + n) • a = m • a + n • a := by
   induction m with
   | zero => rw [Nat.zero_add, zero_nsmul, zero_add]
   | succ m ih => rw [Nat.succ_add, Nat.succ_eq_add_one, succ_nsmul, ih, succ_nsmul, add_assoc]
 
-end AddMonoid
-
-section Monoid
-
-variable [Monoid M]
-
-@[to_additive nsmul_zero, simp] theorem one_pow (n : ℕ) : (1 : M)^n = 1 := by
+@[to_additive nsmul_zero, simp]
+theorem one_pow (n : ℕ) : (1 : M) ^ n = 1 := by
   induction' n with n ih
   · exact pow_zero _
   · rw [pow_succ, ih, one_mul]
 
 @[simp, to_additive one_nsmul]
 theorem pow_one (a : M) : a ^ 1 = a := by rw [pow_succ, pow_zero, mul_one]
 
+/-- Note that most of the lemmas about powers of two refer to it as `sq`. -/
+@[to_additive two_nsmul ""]
+theorem pow_two (a : M) : a ^ 2 = a * a := by rw [pow_succ, pow_one]
+#align pow_two pow_two
+
+alias pow_two ← sq
+
+theorem pow_three' (a : M) : a ^ 3 = a * a * a := by rw [pow_succ', pow_two]
+#align pow_three' pow_three'
+
+theorem pow_three (a : M) : a ^ 3 = a * (a * a) := by rw [pow_succ, pow_two]
+#align pow_three pow_three
+
 @[to_additive add_nsmul]
 theorem pow_add (a : M) (m n : ℕ) : a ^ (m + n) = a ^ m * a ^ n := by
   induction' n with n ih
   · rw [Nat.add_zero, pow_zero, mul_one]
   · rw [pow_succ', ← mul_assoc, ← ih, ← pow_succ', Nat.add_assoc]
 
+@[to_additive mul_nsmul']
 theorem pow_mul (a : M) (m n : ℕ) : a ^ (m * n) = (a ^ m) ^ n := by
   induction' n with n ih
   · rw [Nat.mul_zero, pow_zero, pow_zero]
   · rw [Nat.mul_succ, pow_add, pow_succ', ih]
 
+@[to_additive mul_nsmul]
```

Comment:
Should this be `@[to_additive mul_smul']` and the attribute above should also be changed?

Reply:
I did this, and left a porting note.